src/main/studygate/ui.cljs

```diff
@@ -16,64 +16,57 @@
     (when-not (empty? trimmed-text)
       trimmed-text)))
 
-(defn string-input [component callback {:keys [db/id question/displayname question/options question/value]}]
+(defmulti render-question (fn [question-component props] (:question/type props)))
+
+(defmethod render-question :text [this {:keys [db/id question/displayname question/value question/options]}]
   (dom/div #js {:className "question"}
-           (dom/input #js {:className "effect-17"
-                           :type "text"
-                           :value     (or value "")
-                           :onChange #(let [val (-> % .-target .-value)]
-                                        (callback val))})
-           (dom/label nil displayname)
-           (dom/span #js {:className "focus-border"})))
-
-(defn option-input [component callback {:keys [db/id question/displayname question/value question/options]}]
+    (dom/input #js {:className "effect-17"
+                    :type      "text"
+                    :value     (or value "")
+                    :onChange  #(let [val (-> % .-target .-value)]
+                                  (prim/transact! this `[(api/set-question-value ~{:id id :value val})]))})
```

Comment:
The nice thing about this technique is each question type can have both custom rendering *and* mutations. Right now all you're doing is setting one prop, but I can imagine needing other things.

The other possibility is to use a union query for your questions. That way they could each query for different kinds of data...which can be even more useful.